### ISSUE TITLE
Faust2 and LLVM-6.0.1

### ISSUE THREAD

**Issue by muziker** [kmatheussen/radium#1126]
If you are a subscriber requesting a new feature, please make sure I know who you are.
You do that by emailing me your user name. Feature requests by non-subscribers are
likely to be ignored.

If you are reporting a bug, please use the template below.
The most important part of the template is the step-by-step
recipe. When making a step-by-step recipe, you often include
crucial steps that you wouldn't have thought of otherwise.
It's hard to know what's relevant, so just describe all steps
down to detail. Reporting a bug this way makes the job
easier both for you and me, and we avoid many follow-up
questions.

### 1. Steps to reproduce the problem.

* Step 1: RADIUM_QT5_VERSION=5 make packages

* I expected this to happen: packages compile

* This is what actually happened:
<code>
generator/dsp_aux.o: In function `expandDSPFromFile(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&
, int, char const**, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, std::__cxx11::basic_string<char, st
d::char_traits<char>, std::allocator<char> >&)':
/home/radium/bin/packages/faust2/compiler/generator/dsp_aux.cpp:60: undefined reference to `path_to_content(std::__cxx11::basi
c_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::alloca
tor<char> >&)'
generator/dsp_aux.o: In function `expandDSPFromString(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > cons
t&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int, char const**, std::__cxx11::basic_string<
char, std::char_traits<char>, std::allocator<char> >&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&)':
/home/radium/bin/packages/faust2/compiler/generator/dsp_aux.cpp:78: undefined reference to `gDSPFactoriesLock'
/home/radium/bin/packages/faust2/compiler/generator/dsp_aux.cpp:78: undefined reference to `gDSPFactoriesLock'
generator/dsp_aux.o: In function `generateAuxFilesFromFile(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >
 const&, int, char const**, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&)':
/home/radium/bin/packages/faust2/compiler/generator/dsp_aux.cpp:117: undefined reference to `path_to_content(std::__cxx11::bas
ic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::alloc
ator<char> >&)'
generator/dsp_aux.o: In function `generateAuxFilesFromString(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>
 > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int, char const**, std::__cxx11::basic_
string<char, std::char_traits<char>, std::allocator<char> >&)':
/home/radium/bin/packages/faust2/compiler/generator/dsp_aux.cpp:128: undefined reference to `gDSPFactoriesLock'
/home/radium/bin/packages/faust2/compiler/generator/dsp_aux.cpp:128: undefined reference to `gDSPFactoriesLock'
libcode.o: In function `load_module(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, llvm::LLVMContext*)':
/home/radium/bin/packages/faust2/compiler/libcode.cpp:115: undefined reference to `load_single_module(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, llvm::LLVMContext*)'
/home/radium/bin/packages/faust2/compiler/libcode.cpp:122: undefined reference to `load_single_module(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, llvm::LLVMContext*)'
collect2: error: ld returned 1 exit status
Makefile.unix:227: recipe for target 'faust' failed
make[2]: *** [faust] Error 1
make[2]: Leaving directory '/home/radium/bin/packages/faust2/compiler'
Makefile:44: recipe for target 'all' failed
make[1]: *** [all] Error 2
make[1]: Leaving directory '/home/radium/bin/packages/faust2'
Makefile:543: recipe for target 'bin/packages/deletemetorebuild' failed
</code>

* I've tried: 
Using llvm 4 instead. On a ubuntu system this generates the same error. llvm4.0.0 and llvm6.0.1are installed in separate directories.

### 2. Which version of Radium are you using? Do you know if an earlier version has worked?
git describe -> 5.7.9

### 3. If relevant, which operating system did you run Radium on?
ubuntu low latency kernel

**Comment by muziker** [comment]
I do have faust compiled and running. Is it possible to skip radium/faust2 compile and redirect to the local faust installation ?

**Comment by kmatheussen** [comment]
Which error message did you get when using llvm 4? llvm 4 should be supported.

> I do have faust compiled and running. Is it possible to skip radium/faust2 compile and redirect to the local faust installation ?

Yes, probably, but radium needs custom faust anyway since libfaust and so forth is used for the faustdev instrument. Or, at least, a custom faust makes things much easier to compile up radium, unless faust doesn't compile in the first place of course. However, if you change WHICHFAUST in Makefile, and remove export INCLUDE_FAUSTDEV="jadda" from build_linux_common.sh, it might work to remove faust from bin/packages/build.sh as well.

**Comment by muziker** [comment]
I've recompiled faust2, by going into packages and compiling from there. It seems to work, using LLVM-4.0. I'll keep the other build notes in mind if building with llvm-6.0.1

**Comment by Teteros** [comment]
You're right, LLVM 4.0.1 is the last version of LLVM radium will currently compile with. I've kept radium's faust package up to date with LLVM till upstream diverged too much for simple [backports](https://github.com/kmatheussen/radium/pulls?utf8=%E2%9C%93&q=is%3Apr+LLVM).

Sadly, depending on your graphics driver:
You may experience linking conflicts on runtime when your distro links Mesa (software OpenGL renderer) with LLVM >=4.0.1.
I've discussed this [here](https://users.notam02.no/~kjetism/radium/forum/viewtopic.php?f=7&t=39).

To avoid this, Radium can be build without faustdev support (as Kjetil suggested above) then LLVM4 dependency is only on the `faust` binary, and so this issue is avoided. Though if you're compiling radium and need external llvm40 anyway, you might as well build faust support too.

## Linking LLVM4 statically on Ubuntu

It looks like Ubuntu also uses `/usr/bin/llvm-config-4.0` paths for its [llvm-4.0](https://packages.ubuntu.com/bionic/amd64/llvm-4.0/filelist) and [llvm-4.0-dev](https://packages.ubuntu.com/bionic/amd64/llvm-4.0-dev/filelist) packages (same as LLVM40 AUR package in Arch linux)

So my [patch](https://aur.archlinux.org/cgit/aur.git/tree/use-llvm40-static-libs.patch?h=radium) should also work for Ubuntu.

Quick way to apply it (make sure your current directory is git clone of radium)

```bash
curl -sS "https://aur.archlinux.org/cgit/aur.git/plain/use-llvm40-static-libs.patch?h=radium" | patch -Np1
```
then build radium as usual:
```
RADIUM_QT_VERSION=5 make packages
RADIUM_QT_VERSION=5 BUILDTYPE=RELEASE ./build_linux.sh
```

[This issue is a duplicate of #1068.]
